### ISSUE TITLE
Remote Administration

### ISSUE THREAD

**Issue by bretcope** [NFig/NFig#14]
A clear use case is starting to emerge for allowing application settings the be managed remotely (instead of from within the app that _consumes_ the settings). The reasons for this include:

- Not all apps are web apps (makes it difficult to build a web admin panel).
- Each app with an admin panel has to handle its own authentication.
- Implementing am admin panel is not trivial, and feels like unnecessary work to duplicate (and maintain) across lots of apps.
- Worst case, a bad setting value might prevent access to that app's admin panel (making it difficult to change the bad setting).

Having a single admin panel which can control multiple applications makes sense.

### We _only_ store _overrides_ in Redis  (or whatever the external store is)

Outside of the app, there is no way to know what other settings exist, their default values, or their description/type/etc.

We'll need the app to upload its metadata and default values to the remote store so it can be read by a remote admin panel. This should be relatively straight forward. We can reuse the JSON models being designed for the admin panel (see #9).

### Discoverability

Multiple apps will use the same Redis instance, and the remote admin panel should be able to discover the apps without having them hard-coded. Each app will need to add an entry for itself in a table/set so that the app is discoverable by the remote admin panel.

### Custom Converters

All setting values can be represented a string. This helps normalize how overrides are stored. A simple [ISettingConverter<TValue>](https://github.com/NFig/NFig/blob/master/NFig/SettingConverters.cs#L15) interface is used to convert to and from typed values and their string representation. Default converters exist for all primitive types, but you can override these defaults, and you can build custom converters to support non-primitive types.

Before setting an override, NFig makes sure the string value is able to pass through the converter without throwing an exception. However, there's no way for the remote admin panel to execute a custom converter. Therefore, we can't remotely validate an setting which uses a non-default converter.

In practice, though, custom converters aren't used much. Most settings are just primitives (booleans, numbers, strings) using the default converters, so I think this isn't a big issue. We can simply warn users when they're about to set an override that can't be validated remotely. If an override is invalid, NFig reverts back to the default value, so that mitigates the risk slightly.

### Encrypted Settings

Similar to custom converters, we won't be able to validate overrides for encrypted settings. There's really no way around this.

### Remote Interface

There is no interface for setting/clearing overrides remotely. NFig only understands the settings loaded from code, and won't let you set arbitrarily-named overrides.

I don't think it makes sense to mutilate NFigStore because the use cases are pretty distinct. You either want to act as an admin panel for lots of apps, or you want to be an app consuming settings. Supporting both use cases in one interface might lead to user confusion.

I'm considering the possibility of separating out admin actions entirely from NFigStore. Perhaps renaming it to NFigClient or something like that. Then, all of the admin actions (clear/set overrides, restore snapshot) would be moved to another class (NFigAdmin maybe?), and that class could get its data in a few different ways based on whether the settings are remote or local.

I don't know. I need to mock out an interface and look at the code to see what makes sense. It's possible that approach will create more problems than it solves. I'm just sort of thinking out loud at the moment. Either way, there are lots of breaking changes coming with NFig 3 anyway, so now is a good time to decide how we want this to work.

**Comment by rossipedia** [comment]
To clarify: the idea here is to be able to update settings for a given app (or group of apps) _without_ communicating directly with that app, correct? 

That feels like it would have to be tightly coupled with the backing store implementation (which right now wouldn't be _too_ big of an undertaking as there's really only the Redis one at the moment), as well have the restraint that a given remote admin UI would only be able to manage a set of apps that share the same backing store implementation. 

(ok, gonna brain-dump here for a little bit)

---

As NFig itself is just a library, the concept of "local" vs "remote" would have to exist at the application level, as `NFigStore` has only two responsibilities: read settings and set overrides. Any remote abstraction wouldn't belong in `NFig` itself. 

Specifying those abstractions is probably a good idea though. My initial thoughts here are that there are two primary scenarios:

* Web Apps
* Non-Web Apps

## Web Apps

In a web application, we can provide the UI and endpoints necessary to view settings and set overrides, and the app can host the management UI. This is the current status quo, and looks like this:

```
--------------                       ---------------                       ---------------
|            | <-- Read Settings <-- |             | <-- Read Settings <-- |             |
|  Admin UI  |                       |   Web App   |                       |  NFigStore  |
|            | --> Set Overrides --> |             | --> Set Overrides --> |             |
--------------                       ---------------                       ---------------

```

# Non-Web Apps

In a non-web scenario things are a bit tricky, as the consuming app may be under the restriction that it may not expose any network endpoints. Also, _changing_ a setting value (ie: setting an override) is orthagonal to consuming a given setting. A non-web app would need to consume settings, but not set them. It _would_ however, need to be _notified_ when a new override has been set for it.

Given that the `NFigStore` instance is going to live in each app, we want the flow to look something like this:
```
-----------------                       ---------------
|               | <-- Read Settings <-- |             |
|  Non Web App  |                       |  NFigStore  |
|               | <--- Get Updates <--- |             |
-----------------                       ---------------
                                            |    ▲
                            Store Overrides |    | Read Settings
                                            ▼    |
                                        ---------------                       ---------------
                                        |             | --> Read Settings --> |             |
                                        | Persistence |                       |  Remote UI  |
                                        |             | <-- Set Overrides <-- |             |
                                        ---------------                       ---------------
```

I have some more thoughts on this, but I want to see if I'm on the right track here before I get too deep, heh.

**Comment by bretcope** [comment]
That's more or less accurate. It will be tightly coupled with the store that you choose. You won't be able to have an app using, say, NFig.Sql and the remote admin panel using NFig.Redis. That would be nonsensical.

I do think the base classes for both the consumer side and the admin side belong in NFig. Each store-providing library (e.g. NFig.Redis) will also need to implement both because they will almost certainly share some code. I also just think it would be a bit of a mess to have to have NFig.Client, NFig.Admin, NFig.Client.Redis, NFig.Admin.Redis, etc.

Again though, I'm not 100% sure I want to split out admin vs client. I'm leaning towards it, but I guarantee there are going to be some big pain points. It's going to require more thought and discovery.

**Comment by NickCraver** [comment]
This writeup looks awesome - I'm happy to get Opserver bits in soon as this lands in NuGet pre-release to make sure we're covered and find any issues. Lemme know if I can help design through issues, I'm psyched to finally tackle this problem. Yay NFig!

**Comment by captncraig** [comment]
Have you considered a standalone server with a rest API for administering settings? Might be cleaner than having both sides trying to share the underlying store. Could use long polling or a websocket for notifications. And would make it easier for non-.net clients to get started.

**Comment by captncraig** [comment]
Nvm, I think that's what this is, I'm just confused by block diagrams I think.

**Comment by bretcope** [comment]
The accurate diagram is:

![nfig 3 0 diagram](https://user-images.githubusercontent.com/3608260/26893719-22159dd0-4b8a-11e7-9181-3ca75b73c426.png)

The jagged/dotted line indicates that the consuming app and admin app may or may not actually be the same application.

@captncraig I think what you were suggesting is the possibility of adding an extra layer of abstraction between NFigStore and Redis. It's an interesting idea. The positives would be that we could enforce that the backing store is used consistently by all clients (could be especially helpful for non-.NET clients, like Go), we could introduce another layer of caching to scale Redis reads better, and the interface could be agnostic of the persistence layer.

The negatives are that it's another level of abstraction which would have to be built, set up, and maintained. Performance will be slower. It's another point of failure, another layer of caching problems, and another failover concern.

I'm open to @NickCraver's opinion on the matter, but my initial opinion is that we should stick with direct Redis communication for the first rollout. In part, because it's less work. All of the work I'm doing now is still required either way. The architecture of NFig is such that it would be pretty easy to introduce another layer of abstraction after the fact. We would just write a proxy-store that inherits from NFigStore, and makes HTTP calls instead of Redis calls.

**Comment by NickCraver** [comment]
I agree with @bretcope above - direct redis is the way to go and if we have a REST API it'd be built on top of that anyway. I think that'd be a nice addition to have somewhere (as an example perhaps) later...but totally agreed let's make the direct thing work before working on the API a layer up from it.
